### ISSUE TITLE
Can PyTorch support using custom compilers for model's components and the Inductor compiler for the model?

### ISSUE THREAD

**Issue by taomiao** [pytorch/pytorch#115526]
### 🚀 The feature, motivation and pitch

```
import torch
from torch._dynamo.eval_frame import _debug_get_cache_entry_list

def mycompiler(gm: torch.fx.GraphModule, example_inputs):
    return torch.jit.trace(gm.forward, example_inputs)

def pre_break_graph(model, inputs):
    if torch.sum(inputs[0])>1:
        return inputs

def after_break_graph(model, inputs, outputs):
    if torch.sum(outputs[0])>1:
        return outputs

class Sub1(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 10)
        # self.register_forward_pre_hook(pre_break_graph)
        # self.register_forward_hook(after_break_graph)


    def forward(self, x):
        x = x+1
        return self.linear(x)

class Sub2(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 10)
        # self.register_forward_pre_hook(pre_break_graph)
        # self.register_forward_hook(after_break_graph)

    def forward(self, x):
        x = x+2
        return self.linear(x)

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.sub1 = Sub1()
        self.sub2 = Sub2()

    def forward(self, x):
        x = self.sub1(x)
        x = x + 100
        x = self.sub2(x)
        return x



if __name__=="__main__":
    m = Model()
    for name, sub in m._modules.items():
        if name == "":
            continue
        opt_sub = torch.compile(sub, backend=mycompiler)
        m.register_module(name, opt_sub)

    opt_m = torch.compile(m, backend="inductor")
    res = opt_m(torch.ones(10, 10))
    print(opt_m.forward)
```
code like this does not work.

**Comment by masnesral** [comment]
I'm helping scrub old issues this week. Looks like this issue never got much discussion. I don't know who to tag... maybe @anijain2305. Animesh: would this be a dynamo feature? Is this something we'd ever be interested in supporting?

**Comment by ezyang** [comment]
Assuming you have a way to pattern match for the stuff you're looking for (now that we inline nn modules, you can't directly match on nn modules without some sort of unflatten step), you can write a custom FX pass to swap this out with some custom compiled kernel for your model.  I'm not sure what's going on with their sample code though.

**Comment by taomiao** [comment]
> Assuming you have a way to pattern match for the stuff you're looking for (now that we inline nn modules, you can't directly match on nn modules without some sort of unflatten step), you can write a custom FX pass to swap this out with some custom compiled kernel for your model. I'm not sure what's going on with their sample code though.

What I want to do is use a custom compiler (such as tvm) for part of the model, and the entire nn.Module can also be compiled by the inductor

**Comment by StrongerXi** [comment]
Yeah right now `mycompiler` will never fire in your example because the outer `torch.compile`'s `backend` setting overrides the regional one `opt_sub = torch.compile(sub, backend=mycompiler)`.

> What I want to do is use a custom compiler (such as tvm) for part of the model, and the entire nn.Module can also be compiled by the inductor

I think the closest we have for this is higher order ops (plus some backend compiler annotation and plumbing), but I don't know if there's enough incentive to work on a hop for this specific use case. cc @zou3519

So the workaround is probably as Ed suggested above, using a custom FX pass to pattern match on the part of the model you care about, and then compile that into a custom op or something, and replace the old nodes in the graph into a call to that compiled subgraph.

**Comment by zou3519** [comment]
Answer is still use a custom FX pass.

One alternative is to use a custom Dynamo backend that can split the graph and call Inductor or call a custom compiler.
